File: lovebrew/data/classes/hac.py
import os
from pathlib import Path

from .console import Console


class HAC(Console):

    ENV_PATH = Path(os.getenv("DEVKITPRO"))

    COMMANDS = {
        "meta": "nacptool --create '{name}' '{author}' '{version}' {dst}.nacp",
        "binary": "elf2nro {elf} {dst}.nro --icon={icon} --nacp={dst}.nacp --romfsdir={romfs}"
    }

    def __init__(self, data):
        super().__init__(data)

    def build(self):
        super().build()

        self.build_directory.mkdir(exist_ok=True)

        command = HAC.COMMANDS["meta"].format(name=self.name,
                                              author=self.author,
                                              version=self.version,
                                              dst=self.output_directory / self.target_name.name)

        self._run_command(command)

        command = HAC.COMMANDS["binary"].format(elf=self.get_binary(),
                                                icon=self.get_icon(),
                                                dst=self.output_directory / self.target_name.name,
                                                romfs=self.source_directory.name)

        self._run_command(command)

    def __str__(self):
        return "Nintendo Switch"

File: source/classes/hac.nim
import console
export console

import tables
import strutils

import os

import ../config/configfile
import ../prompts

let meta_cmd = "nacptool --create '$1' '$2' $3 $4.nacp"
let bin_cmd  = "elf2nro $1 $2.nro --icon=$3 --nacp=$2.nacp"

type HAC* = ref object of ConsoleBase

proc getName*(self : HAC) : string = "Nintendo Switch"
proc getBinaryName*(self : HAC) : string = "Switch.elf"
proc getIconExtension*(self: HAC) : string = "jpg"
proc getExtension*(self : HAC) : string = "nro"

proc publish(self : HAC, source : string) : bool =
    let elfBinaryPath = self.getBinaryPath()

    if not elfBinaryPath.fileExists() and not Config.shouldOutputRawData():
        BUILD_FAIL.showFormatted(source, self.getName(), self.getBinaryName(), self.getBinaryPath())
        return false

    # Create metadata
    runCommand(meta_cmd.format(self.name, self.author, self.version, elfBinaryPath))

    # Create binary
    runCommand(bin_cmd.format(self.getBinaryPath(), self.name, self.getIcon()))

    return self.packGameDirectory(source)

when isMainModule:
    import unittest
    test "Name":
        var switch = HAC(name: "my game", author: "me", description: "cool", version: "1.2")
        check switch.getName() == "Nintendo Switch"
        check switch.getProjectName() == "my game"
        check switch.getBinaryName() == "Switch.elf"
        check switch.getIconExtension() == "jpg"
        check switch.getExtension() == "nro"
